sky/global_user_state.py

```diff
@@ -255,12 +326,75 @@ def set_cluster_metadata(cluster_name: str, metadata: Dict[str, Any]) -> None:
         raise ValueError(f'Cluster {cluster_name} not found.')
 
 
+def get_cluster_usage_intervals(cluster_hash: str) -> Optional[Dict[str, Any]]:
+    rows = _DB.cursor.execute(
+        'SELECT usage_intervals FROM cluster_history WHERE cluster_hash=(?)',
+        (cluster_hash,))
+    for (usage_intervals,) in rows:
+        if usage_intervals is None:
+            return None
+        return pickle.loads(usage_intervals)
+
+
+def set_cluster_usage_intervals(cluster_hash: str,
+                                usage_intervals: Dict[str, Any]) -> None:
+    _DB.cursor.execute(
+        'UPDATE cluster_history SET usage_intervals=(?) WHERE cluster_hash=(?)',
+        (
+            pickle.dumps(usage_intervals),
+            cluster_hash,
+        ))
+    count = _DB.cursor.rowcount
+    _DB.conn.commit()
+    assert count <= 1, count
+    if count == 0:
+        raise ValueError(f'Cluster hash {cluster_hash} not found.')
+
+
+def get_cluster_hash(cluster_name: str) -> Optional[str]:
+    rows = _DB.cursor.execute(
+        'SELECT cluster_hash FROM clusters WHERE name=(?)', (cluster_name,))
+    for (cluster_hash,) in rows:
+        if cluster_hash is None:
+            return None
+        return cluster_hash
+
+
+def set_cluster_hash(cluster_name: str, cluster_hash: str) -> None:
+    _DB.cursor.execute('UPDATE clusters SET cluster_hash=(?) WHERE name=(?)', (
+        cluster_hash,
+        cluster_name,
+    ))
+    count = _DB.cursor.rowcount
+    _DB.conn.commit()
+    assert count <= 1, count
+    if count == 0:
+        raise ValueError(f'Cluster {cluster_name} not found.')
+
+
+def get_launched_resources(handle: Optional['backends.Backend.ResourceHandle']):
+    launched_resources = handle.launched_resources
+    launched_nodes = handle.launched_nodes
+    return (launched_nodes, launched_resources)
```

Comment:
This function is used only once. It is better to keep it inline at the place it is called.